crates/ruff_python_ast/src/identifier.rs

```diff
@@ -0,0 +1,587 @@
+        .next()
+        .expect("Failed to find `except` token in `Excepthandler`")
+}
+
+/// Return the [`TextRange`] of the `else` token in a `For`, `AsyncFor`, or `While` statement.
+pub fn else_(stmt: &Stmt, locator: &Locator) -> Option<TextRange> {
+    let (Stmt::For(ast::StmtFor { body, orelse, .. })
+        | Stmt::AsyncFor(ast::StmtAsyncFor { body, orelse, .. })
+        | Stmt::While(ast::StmtWhile { body, orelse, .. })) = stmt else {
+        return None;
+    };
+
+    if orelse.is_empty() {
+        return None;
+    }
+
+    IdentifierTokenizer::starts_at(
+        body.last().expect("Expected body to be non-empty").end(),
+        locator.contents(),
+    )
+    .next()
+}
+
+/// Return `true` if the given character is a valid identifier character.
+fn is_python_identifier(c: char) -> bool {
+    c.is_alphanumeric() || c == '_' || c == '.'
+}
+
+/// Simple zero allocation tokenizer for Python identifiers.
+///
+/// The tokenizer must operate over a range that can only contain identifiers, keywords, and
+/// comments (along with whitespace and continuation characters). It does not support other tokens,
+/// like operators, literals, or delimiters. It also does not differentiate between keywords and
+/// identifiers, treating every valid token as an "identifier".
+///
+/// This is useful for cases like, e.g., identifying the alias name in an aliased import (`bar` in
+/// `import foo as bar`), where we're guaranteed to only have identifiers and keywords in the
+/// relevant range.
+pub(crate) struct IdentifierTokenizer<'a> {
+    cursor: Cursor<'a>,
+    offset: TextSize,
+}
+
+impl<'a> IdentifierTokenizer<'a> {
+    pub(crate) fn new(source: &'a str, range: TextRange) -> Self {
+        Self {
+            cursor: Cursor::new(&source[range]),
+            offset: range.start(),
+        }
+    }
+
+    pub(crate) fn starts_at(offset: TextSize, source: &'a str) -> Self {
+        let range = TextRange::new(offset, source.text_len());
+        Self::new(source, range)
+    }
+
+    fn next_token(&mut self) -> Option<TextRange> {
+        while let Some(c) = self.cursor.bump() {
+            match c {
+                c if is_python_identifier(c) => {
+                    let start = self.offset.add(self.cursor.offset()).sub(c.text_len());
+                    self.cursor.eat_while(is_python_identifier);
+                    let end = self.offset.add(self.cursor.offset());
+                    return Some(TextRange::new(start, end));
+                }
+
+                c if is_python_whitespace(c) => {
+                    self.cursor.eat_while(is_python_whitespace);
+                }
+
+                '#' => {
+                    self.cursor.eat_while(|c| !matches!(c, '\n' | '\r'));
+                }
+
+                '\r' => {
+                    self.cursor.eat_char('\n');
+                }
+
+                '\n' => {
+                    // Nothing to do.
+                }
+
+                '\\' => {
```

Comment:
Is this in the context of escaped newlines?

Reply:
No, these are intended to match continuation characters.

Reply:
yep that's what i meant but didn't have the right term